lambdas/choice/src/writer.py

```diff
@@ -1,6 +1,89 @@
 import os
+import boto3
+
+"""
+Processes SNS messages containing user choices by writing choices to Dynamo.
+For event and database record shapes, see https://github.com/sjawhar/enchanted-brain/issues/4.
+"""
+
+
+EVENT_USER_ID_KEY = 'userId'
+EVENT_CHOICE_TYPE_KEY = 'choiceType'
+EVENT_CHOICE_KEY = 'choice'
+EVENT_TIMESTAMP_KEY = 'timestamp'
+
+CHOICE_COLOR = 'CHOICE_COLOR'
+CHOICE_EMOTION = 'CHOICE_EMOTION'
+CHOICE_CHILLS = 'CHOICE_CHILLS'
+CHOICE_IMAGERY = 'CHOICE_IMAGERY'
+
+CONCERT_SONG_ID = os.environ['EVENT_NAME']
+DYNAMO_DB_TABLE_NAME = os.environ['DYNAMO_TABLE_NAME']
+
+choice_type_to_choice_key = {
+    CHOICE_COLOR: 'colors',
+    CHOICE_EMOTION: 'emotions',
+    CHOICE_CHILLS: 'chills',
+    CHOICE_IMAGERY: 'imagery'
+}
+choice_types = choice_type_to_choice_key.keys()
+
+dynamodb = boto3.resource(
+    'dynamodb',
+    region_name=os.environ['AWS_DYNAMO_REGION'],
+    endpoint_url=os.environ['AWS_DYNAMO_ENDPOINT'],
+    aws_secret_access_key=os.environ['AWS_SECRET_ACCESS_KEY'],
+    aws_access_key_id=os.environ['AWS_ACCESS_KEY_ID']
+)
+table = dynamodb.Table(DYNAMO_DB_TABLE_NAME)
+
 
 def handler(event, context):
-  print(event)
-  print(os.environ)
-  return { 'statusCode': 204 }
+    return write_to_dynamo(dynamodb, event)
+
+
+def write_to_dynamo(table, event):
+
+    # Get data from event
+    listen_id = event[EVENT_USER_ID_KEY]
+    choice_type = event[EVENT_CHOICE_TYPE_KEY]
+    choice = event[EVENT_CHOICE_KEY]
+    timestamp = event[EVENT_TIMESTAMP_KEY]
+
+    # If choice type not recognized, return 422
+    if choice_type not in choice_types:
+        return {'statusCode': 422}
+    # Else get database key for choice type
+    choice_key = choice_type_to_choice_key[choice_type]
+
+    # If imagery choice, just add image word
+    if choice_key == choice_type_to_choice_key[CHOICE_IMAGERY]:
+        choice_value = choice
+        response = table.update_item(
+            Key={
+                'songId': CONCERT_SONG_ID,
+                'listenId': listen_id
+            },
+            UpdateExpression='set {} = :choice_value'.format(choice_key),
+            ExpressionAttributeValues={
+                ':choice_value': choice_value,
+            },
+            ReturnValues='NONE'
+        )
+
+    # Else append timestamp: choice
+    else:
+        choice_value = {timestamp: choice}
+        # If attribute not present, add an empty list, then append to it
+        response = table.update_item(
+            Key={
+                'songId': CONCERT_SONG_ID,
+                'listenId': listen_id
+            },
+            UpdateExpression='SET {} = list_append(if_not_exists({}, :empty_list), :choice_value)'.format(choice_key, choice_key),
```

Comment:
Yep, you're absolutely right. Dumb reading on my part: I saw ellipses in the design doc and my mind went to list of maps rather than just a single map. Will fix.

Reply:
Two follow-up issues here:

1: As I understand it, Dynamo doesn't allow insertion into maps for attributes that do not yet exist. This means that if a row hasn't yet been created for a user, or if an attribute has not yet been created for a given choice type, the update for that choice for that user will fail. Can we expect user rows and attribute columns to be created before invocation of this Lambda? If we don't already have a plan for that, it might make sense to do a conditional put during authentication.

2: Dynamo seems not to like non-String object keys. With that in mind, is it alright if we store timestamps as Strings rather than as Numbers?

Reply:
> Can we expect user rows and attribute columns to be created before invocation of this Lambda?
I was planning to have the record for the user created on connect, with `colors`, `emotions`, etc. initialized as empty maps.

> is it alright if we store timestamps as Strings rather than as Numbers?
Good catch! I don't see a problem with that.